src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php

```diff
@@ -0,0 +1,133 @@
+use Symfony\Component\Serializer\SerializerInterface;
+use Symfony\Component\Validator\Exception\ValidationFailedException;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+
+/**
+ * @author Konstantin Myakshin <molodchick@gmail.com>
+ */
+final class RequestPayloadValueResolver implements ValueResolverInterface
+{
+    private const DEFAULT_FORMAT = 'json';
+
+    private const CONTEXT_DENORMALIZE = [
+        AbstractObjectNormalizer::DISABLE_TYPE_ENFORCEMENT => true,
+        DenormalizerInterface::COLLECT_DENORMALIZATION_ERRORS => true,
+    ];
+
+    private const CONTEXT_DESERIALIZE = [
+        DenormalizerInterface::COLLECT_DENORMALIZATION_ERRORS => true,
+    ];
+
+    public function __construct(
+        private readonly ?SerializerInterface $serializer,
+        private readonly ?DenormalizerInterface $denormalizer,
+        private readonly ?ValidatorInterface $validator,
+    ) {
+    }
+
+    public function resolve(Request $request, ArgumentMetadata $argument): iterable
+    {
+        $payloadMappers = [
+            MapQueryString::class => $this->mapQueryString(...),
+            MapRequestBody::class => $this->mapRequestBody(...),
+        ];
+
+        $payloads = [];
+        foreach ($payloadMappers as $mappingAttribute => $payloadMapper) {
+            $attributes = $argument->getAttributesOfType($mappingAttribute);
+
+            if (!$attributes) {
+                continue;
+            }
+
+            $type = $argument->getType();
+            if (!$type) {
+                throw new \LogicException(sprintf('Could not resolve the "$%s" controller argument: argument should be typed.', $argument->getName()));
+            }
+
+            $payload = $payloadMapper($request, $type, $attributes[0]);
+
+            if ($this->validator) {
+                $violations = $this->validator->validate($payload);
+
+                if (\count($violations)) {
+                    throw new ValidationFailedException($payload, $violations);
+                }
+            }
+
+            $payloads[] = $payload;
+        }
+
+        return $payloads;
+    }
+
+    private function mapQueryString(Request $request, string $type, MapQueryString $attribute): ?object
+    {
+        return $this->denormalize($request->query->all(), $type, $attribute->context);
+    }
+
+    private function mapRequestBody(Request $request, string $type, MapRequestBody $attribute): ?object
+    {
+        if ($request->request->all()) {
+            return $this->denormalize($request->request->all(), $type, $attribute->context);
+        }
+
+        $format = $request->getPreferredFormat(self::DEFAULT_FORMAT);
+
+        try {
+            return $this->deserialize($request->getContent(), $type, $format, $attribute->context);
+        } catch (NotEncodableValueException $exception) {
+            throw new HttpException(Response::HTTP_BAD_REQUEST, sprintf('Request payload contains not valid "%s".', $format), $exception);
+        }
```

Comment:
this is already caught in deserialize(), does it mean we cant detect 415 v 400?

Reply:
indeed, good catch. Yeah, `NotEncodableValueException` throws in both cases: unsupported format, malformed json. We can introduce child exception `UnsupportedFormatException` that extends `NotEncodableValueException`

Reply:
done :heavy_check_mark: